Adds workflow files for CD.

jmchamica opened the pull request:
Workflow:
- Changes pushed to branch **release** or any sub-branch **release/*** will trigger the workflow;
- Project will be built and tested;
- If all tests succeed: create schema on the "development" database; attempt to migrate schema from the "development" database to the main/real database;
- If schema migration succeeds: build a docker image, push to GCP's Container Registry and update the GCE instance.

jmchamica committed:
Adds workflow files for CD.

The workflows are as follows:
- Dev deploy: builds the docker image if the server's source code is
modified to master and publishes it on Cloud run
- Release deploy: builds the docker image when the branch `release` is
updated and publishes it on Cloud run

Both container images will publish to different Cloud run locations.

Further considerations:
- there is no `build/testing` of the project, `assemble` is used instead
to create the .jar file (we might want to add a `testing` step later on)
- Cloud run will not make use of the load balancer, VMs and other
GCP resources configured

pmhsfelix commented on .github/workflows/gcp_dev_deploy.yaml:
```diff
@@ -0,0 +1,64 @@
+        working-directory: project
+        run: docker push $GCP_REPO/$PROJECT_ID/$IMAGE_NAME:$GITHUB_SHA
+
+      - name: Deploy to cloud run
```
What is the goal for this PR. I was assuming that we are not going to use cloud run, and use a VM with an LB instead.

pmhsfelix commented on .github/workflows/gcp_dev_deploy.yaml:
```diff
@@ -0,0 +1,64 @@
+env:
+  PROJECT_ID: ${{ secrets.GCP_PROJ_ID }}
+  RUN_REGION: europe-west1
+  IMAGE_NAME: release-img
```
This should include the `core` name.

jmchamica committed:
[Deploy]: Workflow for GCE deploy.

jmchamica committed:
[Deploy]: Remove dev deploy workflow files (for now)

jmchamica committed:
[Deploy]: Include Sub-branches of "release" for deployment.

jmchamica committed:
[Deploy]: Changes Dockerfile. Do not use glob when moving .jar

jmchamica committed:
[Deploy]: Require tests before deployment

jmchamica committed:
GH Actions Workflow. Test workflow will not run on the release branch.

jmchamica committed:
[Deploy]: Schema diff Scripts+Workflow

jmchamica committed:
[Deploy]: Provoke an error on the Schema Diff phase.

This will abort the Github Actions' Job for deployment. Prevents the
delivery of a new server container if the database's schema fails to
adapt to its new version.

jmchamica committed:
[Deploy]: Provoke exit error on Schema Diff's script

jmchamica committed:
[Deploy]: Concludes the step "Schema Migration Attempt" before Deployment.

jmchamica commented on project/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM openjdk:8-jdk-alpine
+
+RUN addgroup -S spring && adduser -S spring -G spring
+EXPOSE 8080/tcp
```
[Reminder]: forgot `USER..`

jmchamica commented on project/src/test/resources/schema_diff/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM alpine
```
[Reminder]: Target specific version

jmchamica committed:
[Deploy]: Removes unsafe schema diff
